### PR TITLE
NEW: API setup update extrafields from name, elementtype and json

### PR DESCRIPTION
**Pull request by JonBendtsen**
NEW setup API can update new extrafields
Third step in fixing #29249 though this pull request can only update.

### CODE REVIEW

**Comment by JonBendtsen** [comment]
This is the json I've tested it with, and it ignores the fields in the json that is not specified in the function

```
	public function updateExtraField($attrname, $label, $type, $pos, $size, $elementtype, $unique = 0, $required = 0, $default_value = '', $param = '', $alwayseditable = 0, $perms = '', $list = '-1', $help = '', $computed = '', $entity = '', $langfile = '', $enabled = '1', $totalizable = 0, $printable = 0, $moreparams = array())
```

```
{       "id": "5",
      "type": "varchar",
      "label": "schone",
      "size": "255",
      "elementtype": "commande",
      "default": null,
      "computed": null,
      "unique": "1",
      "required": "0",
      "param": {
        "options": {
          "": null
        }
      },
      "pos": "107",
      "alwayseditable": "1",
      "perms": null,
      "list": "1",
      "entity": "1",
      "printable": "1",
      "totalizable": "1",
      "langs": null,
      "help": "test, bare en simpel test",
      "css": null,
      "cssview": null,
      "csslist": null,
      "fk_user_author": "2",
      "fk_user_modif": "3"
    }
```

**Comment by JonBendtsen** [comment]
request url `http://localhost/api/index.php/setup/extrafields/commande/one`

This what it looks like in my browser. I decided to use this setup because then it looks more like the delete, and I think it fits better with the style that I see when using the existing method GET extrafields.

API_setup_update_extrafield

**Comment by JonBendtsen** [comment]
@eldy  I think that the conflicts are fixed now, as for the POST I would like to re-make it in the style of delete and update, meaning attrname and elementtype in the url, the rest in json in the request_data

**Review by eldy — COMMENTED** [review]
For name of column or tables db->sanitize is better then db->escape

File: htdocs/core/class/extrafields.class.php [hunk 1]
```diff
@@ -617,19 +690,23 @@ public function update($attrname, $label, $type, $length, $elementtype, $unique
...
 			if ($result > 0 || $type == 'separate') {
 				if ($label) {
+					dol_syslog(get_class($this).'::update_label', LOG_DEBUG);
 					$result = $this->update_label($attrname, $label, $type, $length, $elementtype, $unique, $required, $pos, $param, $alwayseditable, $perms, $list, $help, $default, $computed, $entity, $langfile, $enabled, $totalizable, $printable, $moreparams);
 				}
 				if ($result > 0) {
 					$sql = '';
 					if ($unique) {
-						$sql = "ALTER TABLE ".$this->db->prefix().$table." ADD UNIQUE INDEX uk_".$table."_".$attrname." (".$attrname.")";
+						dol_syslog(get_class($this).'::update_unique', LOG_DEBUG);
```
[the remark this answers is not in the document]
Comment (JonBendtsen) [reply htdocs/core/class/extrafields.class.php:705]: new commit should fix this

**Comment by JonBendtsen** [comment]
@eldy should be sanitized now, and that one phan error I do not understand because it is not in a file I changed?

**Comment by hregis** [comment]
@JonBendtsen replace just 

"* @param	string				$useempty....."

with 

"* @param	int				$useempty...."

**Comment by JonBendtsen** [comment]
@hregis  for the phan error in a file I did not change?

**Comment by hregis** [comment]
@JonBendtsen phan check is completely alien to me! I don't know how it decides to check a file at one time and not at another... ;-)

https://github.com/Dolibarr/dolibarr/pull/29283

**Comment by hregis** [comment]
@JonBendtsen maybe @mdeweerd can answer this question ! ;-)

**Comment by JonBendtsen** [comment]
At the time of posting this comment the contents of the changes to allow for updating has been copied to #29270 where there is also the additional creation of extra fields.

**Comment by hregis** [comment]
@JonBendtsen @eldy just waiting integration of https://github.com/Dolibarr/dolibarr/pull/29283 and after merge develop branch in your PR branch

**Comment by JonBendtsen** [comment]
> @JonBendtsen @eldy just waiting integration of #29283 and after merge develop branch in your PR branch

@hregis  okay, I'll wait for #29283 too and then merge develop into my branch.

(Do you guys have a chat where you're hanging out since you have that kind of information?)

**Comment by hregis** [comment]
@JonBendtsen there must be a conversation channel but I don't remember which one... I'm old school and all this is beyond me and I don't have the time to take the time anymore... but I can help you on github if you put me in the "@hregis" dialog

**Comment by JonBendtsen** [comment]
> @JonBendtsen there must be a conversation channel but I don't remember which one... I'm old school and all this is beyond me and I don't have the time to take the time anymore... but I can help you on github if you put me in the "@hregis" dialog

@hregis  it is probably in French which I don't understand, so it would probably not be that helpful for me

**Comment by mdeweerd** [comment]
> @JonBendtsen phan check is completely alien to me! I don't know how it decides to check a file at one time and not at another... ;-)
> 
> #29283

A far as I see the integration of the commit [3d2a018](https://github.com/Dolibarr/dolibarr/pull/29273/commits/3d2a018342d97a6e0b09b40181de20c5c0a36fc3) brought the following phpdoc that I think is now fixed.

```php
	 * @param	string				$useempty		1=Add an empty value in list, 2=Add an empty value in list only if there is more than 2 entries.
```

So I probably was invalid the develop branch as well.  In that case the issue in the develop branch propagates to the PR.

**Comment by MarcVJ** [comment]
Hello,
I noticed an error on this SQL query : the syntax "IF EXISTS" is not valid in MySQL, this error pops in my log : 
"DoliDBMysqli::query SQL Error message: DB_ERROR_SYNTAX You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near 'IF EXISTS uk_product_extrafields_prixaukgoulitre' at line 1"

I did a grep on the entire dolibarr-20.0.2 file and the only line I found containing "IF EXISTS" is this one in htdocs/core/class/extrafields.class.php : 
$sql = "ALTER TABLE ".$this->db->prefix().$table." DROP INDEX IF EXISTS uk_".$table."_".$attrname;

**Comment by mdeweerd** [comment]
```
htdocs/core/class/extrafields.class.php:776:69:                      $sql = "ALTER TABLE ".$this->db->prefix().$table." DROP INDEX IF EXISTS uk_".$table."_".$this->db->sanitize($attrname);
htdocs/core/class/utils.class.php:1195:33:               fwrite($handle, "DROP TABLE IF EXISTS `".$table."`;\n"); // Dropping table if exists prior to re create it
htdocs/core/class/utils.class.php:1262:35:        fwrite($handle,"DROP PROCEDURE IF EXISTS '$name'.'$proc'$$\n");
```

When I look at the code for the first occurence which matches your error message, the "DROP INDEX" is supposed to undo the create index - however the `"ALTER TABLE ".$this->db->prefix().$table.` is incorrect.  The index name is unique itself, and "DROP INDEX" does not use "ALTER TABLE".  The 'IF EXISTS' seems correct.

**Comment by JonBendtsen** [comment]
@MarcVJ  do you get a different result if you edit the line a little?

from:
`$sql = "ALTER TABLE ".$this->db->prefix().$table." DROP INDEX IF EXISTS uk_".$table."_".$attrname;`

to:
```$sql = "ALTER TABLE ".$this->db->prefix().$table." DROP INDEX IF EXISTS uk_".$table."_".$this->db->sanitize($attrname);```

**Comment by mdeweerd** [comment]
I think it has to be:
```
$sql = "DROP INDEX IF EXISTS uk_".$table."_".$attrname;
```

**Comment by JonBendtsen** [comment]
@mdeweerd  don't forget `.$this->db->sanitize($attrname);`

**Comment by MarcVJ** [comment]
@JonBendtsen I tried with the edited line : 
`$sql = "ALTER TABLE ".$this->db->prefix().$table." DROP INDEX IF EXISTS uk_".$table."_".$this->db->sanitize($attrname); `
 I didn't get the same error message, I created an extra field (with the Product module), tried to create another with the same name, then suppressed it. I got this in the logs : 
```
2025-01-29 08:17:16 ERR     2a01:cb04:6d2:e00:b5c2:233c:cf81:9977 DoliDBMysqli::query SQL Error query: ALTER TABLE llx_product_extrafields ADD test2 varchar(255) NULL
2025-01-29 08:17:16 ERR     2a01:cb04:6d2:e00:b5c2:233c:cf81:9977 DoliDBMysqli::query SQL Error message: DB_ERROR_COLUMN_ALREADY_EXISTS Duplicate column name 'test2'
2025-01-29 08:17:16 ERR     2a01:cb04:6d2:e00:b5c2:233c:cf81:9977 DoliDBMysqli::query SQL Error query: INSERT INTO llx_extrafields( name, label, type, pos, size, entity, elementtype, fieldunique, fieldrequired, param, alwayseditable, perms, langs, list, printable, fielddefault, fieldcomputed, fk_user_author, fk_user_modif, datec, enabled, help, totalizable, css, csslist, cssview ) VALUES('test2', 'Test 2', 'varchar', 100, '255', 1, 'product', 0, 0, 'a:1:{s:7:\"options\";a:1:{s:0:\"\";N;}}', 1, null, null, '1', '0', null, null, 25, 25,'2025-01-29 09:17:16', '1', null, FALSE, null, null, null)
2025-01-29 08:17:16 ERR     2a01:cb04:6d2:e00:b5c2:233c:cf81:9977 DoliDBMysqli::query SQL Error message: DB_ERROR_RECORD_ALREADY_EXISTS Duplicate entry 'test2-1-product' for key 'llx_extrafields.uk_extrafields_name'

```

**Comment by MarcVJ** [comment]
@mdeweerd I also tried : 
`$sql = "DROP INDEX IF EXISTS uk_".$table."_".$this->db->sanitize($attrname);`
An got those logs : 
```
025-01-29 08:25:08 ERR     2a01:cb04:6d2:e00:b5c2:233c:cf81:9977 DoliDBMysqli::query SQL Error query: ALTER TABLE llx_product_extrafields ADD test2 varchar(255) NULL
2025-01-29 08:25:08 ERR     2a01:cb04:6d2:e00:b5c2:233c:cf81:9977 DoliDBMysqli::query SQL Error message: DB_ERROR_COLUMN_ALREADY_EXISTS Duplicate column name 'test2'
2025-01-29 08:25:08 ERR     2a01:cb04:6d2:e00:b5c2:233c:cf81:9977 DoliDBMysqli::query SQL Error query: INSERT INTO llx_extrafields( name, label, type, pos, size, entity, elementtype, fieldunique, fieldrequired, param, alwayseditable, perms, langs, list, printable, fielddefault, fieldcomputed, fk_user_author, fk_user_modif, datec, enabled, help, totalizable, css, csslist, cssview ) VALUES('test2', 'Test 2', 'varchar', 100, '255', 1, 'product', 0, 0, 'a:1:{s:7:\"options\";a:1:{s:0:\"\";N;}}', 1, null, null, '1', '0', null, null, 25, 25,'2025-01-29 09:25:08', '1', null, FALSE, null, null, null)
2025-01-29 08:25:08 ERR     2a01:cb04:6d2:e00:b5c2:233c:cf81:9977 DoliDBMysqli::query SQL Error message: DB_ERROR_RECORD_ALREADY_EXISTS Duplicate entry 'test2-1-product' for key 'llx_extrafields.uk_extrafields_name'
2025-01-29 08:25:31 ERR     2a01:cb04:6d2:e00:b5c2:233c:cf81:9977 DoliDBMysqli::query SQL Error query: DROP INDEX IF EXISTS uk_product_extrafields_test3
2025-01-29 08:25:31 ERR     2a01:cb04:6d2:e00:b5c2:233c:cf81:9977 DoliDBMysqli::query SQL Error message: DB_ERROR_SYNTAX You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near 'IF EXISTS uk_product_extrafields_test3' at line 1

```

**Comment by hregis** [comment]
@JonBendtsen @MarcVJ it doesn't exists ! `DROP INDEX IF EXISTS`

**Comment by hregis** [comment]
https://dev.mysql.com/doc/refman/8.4/en/drop-index.html

**Comment by hregis** [comment]
i send a fix : https://github.com/Dolibarr/dolibarr/pull/32848

**Comment by mdeweerd** [comment]
This resource suggests that `DROP INDEX IF EXISTS index_name ON table_name;` exists.
https://www.tutorialspoint.com/sql/sql-drop-index.htm#drop_index_if_exists

However I did not look in detail and therefore the command should rather ba:
```
$sql = "DROP INDEX IF EXISTS uk_".$table."_".$this->db->sanitize($attrname)." ON ".$this->db->prefix().$table;
```

Maybe that does not exist for the flavor (the error was a syntax error without the ' ON ').

**Comment by hregis** [comment]
@mdeweerd ok for postgresql and maybe for SQL server, but not for Mysql/MariaDB !

**Comment by hregis** [comment]
@mdeweerd https://dev.mysql.com/doc/refman/8.4/en/drop-index.html

**Comment by hregis** [comment]
@eldy @mdeweerd ok sorry, look, it's ok for Postgresql and MariaDB,  but not for Mysql !

https://www.jooq.org/doc/latest/manual/sql-building/ddl-statements/drop-statement/drop-index-statement/drop-index-if-exists/

**Pull request merged**